waiter/src/waiter/scheduler.clj

```diff
@@ -395,50 +396,115 @@
+
+                  state-query-chan
+                  ([{:keys [response-chan service-id]}]
+                    (if service-id
+                      (let [scheduler-state (service-id->state scheduler service-id)
+                            unhealthy-instances (get service-id->unhealthy-instances service-id)
+                            selected-current-state {:unhealthy-instances unhealthy-instances
+                                                    :tracked-failed-instances (get service-id->tracked-failed-instances service-id)
+                                                    :instance-id->failed-health-checks (filter #(contains? (set unhealthy-instances) (key %)) instance-id->failed-health-checks)}]
+                        (async/>! response-chan (merge selected-current-state scheduler-state)))
+                      (async/>! response-chan (merge current-state (state scheduler))))
+                    current-state)
+
+                  timeout-chan
+                  ([]
+                    (timers/start-stop-time!
+                      (metrics/waiter-timer "state" "scheduler-sync")
+                      (let [^DateTime request-apps-time (t/now)
+                            timing-message-fn (fn [] (let [^DateTime now (t/now)]
+                                                       (str "scheduler-syncer: sync took " (- (.getMillis now) (.getMillis request-apps-time)) " ms")))]
+                        (log/trace "scheduler-syncer: querying scheduler")
+                        (if-let [service->service-instances (timers/start-stop-time!
+                                                              (metrics/waiter-timer "core" "scheduler" "app->available-tasks")
+                                                              (do-health-checks (request-available-waiter-apps scheduler)
+                                                                                (fn available [instance health-check-path]
+                                                                                  (available? instance health-check-path http-client))
+                                                                                service-id->service-description-fn))]
+                          (let [available-services (keys service->service-instances)
+                                available-service-ids (map :id available-services)]
+                            (log/debug "scheduler-syncer:" (count service->service-instances) "available services:" available-service-ids)
+                            (doseq [service available-services]
+                              (when (zero? (reduce + 0 (filter number? (vals (select-keys (:task-stats service) [:staged :running :healthy :unhealthy])))))
+                                (log/info "scheduler-syncer:" (:id service) "has no live instances!" (:task-stats service))))
+                            (loop [service-id->unhealthy-instances' {}
```

Comment:
Consider one map, instead of separate maps (something like `id->health-check-context`)